src/engine/strat_engine/backstore/metadata.rs

```diff
@@ -267,70 +264,114 @@ impl StaticHeader {
+                            (None, None) => Ok(None),
+                            (Some(loc_1), None) => {
+                                // Copy 1 has valid Stratis BDA, copy 2 has no magic, re-write copy 2
+                                BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
+                                Ok(Some(loc_1))
+                            }
+                            (None, Some(loc_2)) => {
+                                // Copy 2 has valid Stratis BDA, copy 1 has no magic, re-write copy 1
+                                BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
+                                Ok(Some(loc_2))
+                            }
+                        }
+                    }
+                    (Ok(loc_1), Err(loc_2)) => {
+                        // Re-write copy 2
+                        if loc_1.is_some() {
                             BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
-                            Ok(Some(loc_1))
+                            Ok(loc_1)
                         } else {
-                            BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
-                            Ok(Some(loc_2))
+                            // Location 1 doesn't have a signature, but location 2 did, but it got an error,
+                            // lets return the error instead as this appears to be a stratis device that
+                            // has gotten in a bad state.
+                            Err(loc_2)
                         }
                     }
-                    (None, None) => Ok(None),
-                    (Some(loc_1), None) => {
-                        // Copy 1 has valid Stratis BDA, copy 2 has no magic, re-write copy 2
-                        BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
-                        Ok(Some(loc_1))
+                    (Err(loc_1), Ok(loc_2)) => {
+                        // Re-write copy 1
+                        if loc_2.is_some() {
+                            BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
+                            Ok(loc_2)
+                        } else {
+                            // Location 2 doesn't have a signature, but location 1 did, but it got an error,
+                            // lets return the error instead as this appears to be a stratis device that
+                            // has gotten in a bad state.
+                            Err(loc_1)
+                        }
                     }
-                    (None, Some(loc_2)) => {
-                        // Copy 2 has valid Stratis BDA, copy 1 has no magic, re-write copy 1
-                        BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
-                        Ok(Some(loc_2))
+                    (Err(_), Err(_)) => {
+                        let err_str =
+                            "Appeared to be a Stratis device, but no valid sigblock found";
```

Comment:
Might as well enhance this error message. The contents of both buffers are available, so they might as well be included.

Reply:
New issue: https://github.com/stratis-storage/stratisd/issues/1444